### PR TITLE
make uv_translate_sys_error public

### PR DESCRIPTION
**Pull request by plafer**
References #79.

### CODE REVIEW

File: include/uv-win.h [hunk 1]
```diff
@@ -78,6 +78,7 @@ typedef intptr_t ssize_t;
 # define SIGABRT_COMPAT       6
 #endif
 
+UV_EXTERN int uv_translate_win32_error(int sys_errno);
```
Comment (llafuente) [inline include/uv-win.h]: this is not the place for `UV_EXTERN`

both function should be called the same for example: `uv_translate_os_error`
`UV_EXTERN uv_translate_os_error` only on `uv.h`

Comment (plafer) [inline include/uv-win.h]: Since this a windows only function, I assume the correct way would be to put it in uv.h and wrap it in `#if defined(_WIN32) ... #endif`?

As for the naming of the functions, it was suggested here: https://github.com/libuv/libuv/issues/79#issuecomment-68108984.

Comment (saghul) [inline include/uv-win.h]: I remember that, but we don't have a mapping for Windwos posix style errnos, so I think i'd go with `uv_translate_sys_error` at the moment.

File: src/uv-common.c [hunk 2]
```diff
@@ -155,6 +155,82 @@ static const char* uv__unknown_err_code(int err) {
   return copy != NULL ? copy : "Unknown system error";
 }
 
+int uv_translate_posix_error(int sys_errno) {
+  switch (sys_errno) {
```
Comment (saghul) [inline src/uv-common.c]: you don't need to do all this, on Unix just return -sys_errno, since that's how we map errors.

Comment (plafer) [inline src/uv-common.c]: I did it this way to that if the function is passed something that doesn't make sense in libuv (e.g. ECHILD), then it will return UV_UNKNOWN instead of some garbage value. Is that a non-issue?

The other reason was that the function written this way would also work on Windows. My understanding is that it needed to work on both Unix and Windows, and on Windows it would need to be written like this anyways (unless I'm missing something?), so might as well just write it once for both.

Comment (saghul) [inline src/uv-common.c]: If there is an errno we forgot to map, then we need to map it :-) As for the implementation, even if we go with a single function, it will have 2 implementations, the Unix side, which returns the negated errno, and the Windows one, which translates win32 errors into uv errors.

Comment (plafer) [inline src/uv-common.c]: This is what I'm planning to change:

- Drop the names `uv_translate_posix_error` and `uv_translate_win32_error`.
- Expose `uv_translate_sys_error`
  - Unix implementation: negated errno
  - Windows implementation: the same as the current one

Does this sound right?

Comment (saghul) [inline src/uv-common.c]: Yep, that sounds right.

**Comment by saghul** [comment]
@piscisaureus can you PTAL?  Other @libuv/collaborators are also welcome, of course.

**Comment by saghul** [comment]
@indutny see #79

**Comment by indutny** [comment]
@saghul I do get your point, but the commit message should describe it too (IMO), and there should be a documentation for it.

**Comment by saghul** [comment]
> but the commit message should describe it too (IMO), and there should be a documentation for it.

Agreed! @plafer can you please take care of these?

**Comment by plafer** [comment]
Of course!

File: docs/src/errors.rst [hunk 3]
```diff
@@ -332,3 +332,11 @@ API
...
     Returns the error name for the given error code.  Leaks a few bytes
     of memory when you call it with an unknown error code.
+
+.. c:function:: int uv_translate_sys_error(int sys_errno)
+
+   Returns the libuv error code equivalent to the given platform dependent error
+   code: POSIX error codes on Unix (the ones stored in `errno`), and Win32 error
+   codes on Windows (those returned by `GetLastError()` or `WSAGetLastError()`).
+
+   If `sys_errno` is already a libuv error, it is simply returned.
```
Comment (saghul) [inline docs/src/errors.rst]: please a a ``.. versionchanged::`` block.

File: src/unix/core.c [hunk 4]
```diff
@@ -1218,3 +1218,8 @@ void uv_os_free_passwd(uv_passwd_t* pwd) {
 int uv_os_get_passwd(uv_passwd_t* pwd) {
   return uv__getpwuid_r(pwd);
 }
+
+int uv_translate_sys_error(int sys_errno) {
```
Comment (saghul) [inline src/unix/core.c]: leave an extra newline here (it's 2 between functions)

**Comment by piscisaureus** [comment]
So this patch has now changed to: "make uv_translate_sys_error public" ?

I'm okay with that. I guess I'd personally prefer to have stable error numbers instead of using the negated platform error code on unix-ish platforms, but that seems out of scope for this discussion.

Also, I find calling the function that translates windows error codes to uv error codes `uv_translate_win32_error` really not a bad.

**Comment by saghul** [comment]
@piscisaureus 

> So this patch has now changed to: "make uv_translate_sys_error public" ?

Thas was the initial request in #79.

> Also, I find calling the function that translates windows error codes to uv error codes uv_translate_win32_error really not a bad.

So, back in the day you suggested we need 2 functions because some Windows functions can also return POSIX style errnos.  This patch doesn't consider those, also libuv doesn't have a translation mechanism for those per se, since we prefer Win32 or NT APIs, right?

So, do you have any objections to the current implementation proposal?  I wanted to double-check with you, since you were the one who originally proposed we have 2 functions.

Thanks Bert!

**Comment by cjihrig** [comment]
If it's going to be a public API, there should probably be at least one test, just for sanity.

**Comment by saghul** [comment]
> If it's going to be a public API, there should probably be at least one test, just for sanity.

Indeed!  A couple of tests would be nice: using a real errno and a UV errno, both on Windows an Unix.

**Comment by piscisaureus** [comment]
@saghul 

I do prefer to have two different functions, but I don't feel that strongly about it.
I'm also okay with just exposing `uv_translate_sys_error`, which is what this PR seems to do.

TL;DR ship it.

**Comment by saghul** [comment]
@piscisaureus Thanks for chiming in!

File: test/test-error.c [hunk 5]
```diff
@@ -48,3 +48,23 @@ TEST_IMPL(error_message) {
 
   return 0;
 }
+
+
+TEST_IMPL(sys_error) {
+
+#if defined(_WIN32)
+
```
Comment (cjihrig) [inline test/test-error.c]: Can you remove the line breaks before and after the assertions.

File: test/test-error.c [hunk 6]
```diff
@@ -48,3 +48,23 @@ TEST_IMPL(error_message) {
...
+  ASSERT(uv_translate_sys_error(ERROR_NOACCESS) == UV_EACCES);
+  ASSERT(uv_translate_sys_error(WSAEADDRINUSE) == UV_EADDRINUSE);
+  ASSERT(uv_translate_sys_error(ERROR_BAD_PIPE) == UV_EPIPE);
+
+#else
+
+  ASSERT(uv_translate_sys_error(EPERM) == UV_EPERM);
```
Comment (cjihrig) [inline test/test-error.c]: Can you test passing in existing `UV_` errors and 0 to verify that they come back.

**Comment by cjihrig** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/146/

**Comment by plafer** [comment]
Could someone point me in the right direction as to why this might have failed on some platforms but not others?

**Comment by cjihrig** [comment]
It doesn't look like the test from this PR actually failed anywhere.

File: docs/src/errors.rst [hunk 7]
```diff
@@ -332,3 +332,14 @@ API
...
+
+   .. versionchanged:: 2.0.0 function declared public and UNIX implementation
```
Comment (saghul) [inline docs/src/errors.rst]: it has both unix and windows implementations, so no need to mention that

Comment (saghul) [inline docs/src/errors.rst]: Also, target it at 1.11, let's have it now.

**Comment by cjihrig** [comment]
@saghul does this LGTY now?

**Comment by saghul** [comment]
@cjihrig Not yet, see https://github.com/libuv/libuv/pull/1060#discussion_r80433463

**Comment by cjihrig** [comment]
It looks like those are addressed now. I think comments left using GitHub's code review feature remain even after being addressed.

**Comment by plafer** [comment]
Is there still something missing? If so, I can fix it.

**Comment by saghul** [comment]
Ah, I see now. It's fixed indeed. Now, this PR mentions this API was added in 1.11.0, but we haven't released 1.10 yet. Hopefully that happens this week, so we could add this one in. Thoughts @cjihrig? I'm ok with landing it now, we can fix the version changed when merging.

**Comment by cjihrig** [comment]
Sounds like a fine plan to me. I wonder if libuv should adopt something like the [REPLACEME](https://github.com/nodejs/node/blob/master/doc/releases.md#step-3-update-any-replaceme-tags-in-the-docs) that Node uses for setting the version added during releases.

**Comment by saghul** [comment]
@cjihrig Yeah, we should probably do that.

**Comment by cjihrig** [comment]
I've opened https://github.com/libuv/libuv-release-tool/issues/7, to track adding REPLACEME, but can get this landed for 1.10 after a new CI run.

**Comment by cjihrig** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/157/

**Comment by cjihrig** [comment]
Unrelated to this change, but that might be the greenest CI run I've seen for libuv.

**Comment by saghul** [comment]
LGTM again :-)

**Comment by cjihrig** [comment]
Thanks! Landed in f1863dae6bb27e31644307cf0905d42bd41251c9.
